### ISSUE TITLE
Parallelizing email delivery

### ISSUE THREAD

**Issue by skosch** [gen-smtp/gen_smtp#116]
Hi,

I am using the fantastic gen_smtp via kamilc's mailman Elixir package and I'm trying to send a few hundred emails at once. It so happens that the university's mail server, before sending each email, checks for the recipient domain's DNS MX record. If that fails, it complains ("550 5.4.3 Policy analysis found DNS error on the target domain") and that causes an `{error, no_more_hosts, ...}` with the result that any valid emails sent immediately after don't make it through either.

Assuming everything goes smoothly, I can actually push lots of emails out quickly by `spawn`ing a ton of simultaneous `send` calls, though that method is vulnerable to the issue described above. So: how can I send lots of emails as quickly as possible, and still be immune to `no_more_hosts`-type errors? (I assume this is related to #67 as well?)

I would be grateful if someone could fill this knowledge gap of mine!

**Comment by seriyps** [comment]
> with the result that any valid emails sent immediately after don't make it through either.

Sounds strange. Can't figure out why this may happen. At least, each `gen_smtp_client:send*` call creates new TCP connection to SMTP server, so, one send shouldn't affect any other.

**Comment by skosch** [comment]
That was my assumption, but since I know next to nothing about how Erlang manages connections I'm not sure what to make of it either. I will do some more testing and get back to you.

**Comment by skosch** [comment]
I have to apologize, I was wrong! Lots of testing eventually brought up a whole bunch of 
```
{:error, :retries_exceeded, {:network_failure, "<my mailserver's url>", {:error, :timeout}}}
```
Maybe it really was just a coincidence that this often happened when some of the addresses were invalid?

Internet is working fine, so I suspect the mail server simply doesn't like me trying to make hundreds of connections at once. I will have to find a way to rate-limit my deliveries. Thanks and sorry about the false alarm!

**Comment by mworrell** [comment]
Just adding some notes.  In Zotonic (http://zotonic.com) we use gen_smtp for all outgoing emails. In some sites we have an average of around a 300K emails per month.

You need to limit the number of parallel connections per receiving email server. For this we use a separate gen_server process that tracks which emails are "in flight" and how many per smtp server.

In this way we throttle the outgoing rate of messages and still maintain high throughput overall. 

All emails are stored in mnesia, so we can also handle retries on non fatal errors (4xx or connection issues).

**Comment by Vagabond** [comment]
Honestly, we should probably rewrite the SMTP client to be able to send more than one email per connection and return a list of success/failure for each message. I never got around to it though.

**Comment by skosch** [comment]
@mworrell @Vagabond thanks for chiming in! I was thinking of using poolboy for that purpose. I'll report back. Of course, multiple emails per connection would be even better!

**Comment by Vagabond** [comment]
if we modeled the client as 2 state machines, one for *establishing* the connection and one for *sending* email, it would be pretty easy. We'd just have to refactor things (which is long overdue anyway). It'd also be good to add some kind of a 'trace' to the client, so you could see what is happening without uncommenting those terrible io:formats.

Sadly, I'm unlikely to have the time to do it anytime soon.

**Comment by skosch** [comment]
> Sadly, I'm unlikely to have the time to do it anytime soon.

I wonder if it might be possible to somehow entice you to reconsider :) Or maybe with your guidance I can try this myself? (Only problem is that although I know Elixir well, Erlang is new territory for me.)

This would be a fantastic feature for enterprise use cases where users have access to an existing relay for outgoing mail.

**Comment by Vagabond** [comment]
I will try to take a look at it, but I'll have to see when I can find the time.

**Comment by Vagabond** [comment]
OK, I made a thing:

```
1> gen_smtp_client:send_blocking([{"whatever@test.com", ["andrew@hijacked.us"],  "Subject: testing\r\nFrom: Andrew Thompson \r\nTo: Some Dude \r\n\r\nThis is the email body"}, {"whatever@test.com", ["andrew@hijacked.us"],  "Subject: testing2222\r\nFrom: Andrew Thompson \r\nTo: Some Dude \r\n\r\nThis is the email body"}],  [{relay, "smtp.gmail.com"}, {username, "me@gmail.com"}, {port, 587}, {password, "password"}, {trace_fun, fun io:format/2}]).
MX records for smtp.gmail.com are [{10,"74.125.28.109"},{10,"74.125.28.108"}]
connected to 74.125.28.109; banner was  smtp.gmail.com ESMTP a25sm2423382pgd.26 - gsmtp

Extensions are [{<<"SIZE">>,<<"35882577">>},
                {<<"8BITMIME">>,true},
                {<<"STARTTLS">>,true},
                {<<"ENHANCEDSTATUSCODES">>,true},
                {<<"PIPELINING">>,true},
                {<<"CHUNKING">>,true},
                {<<"SMTPUTF8">>,true}]
Starting TLS
TLS started
Extensions are [{<<"SIZE">>,<<"35882577">>},
                {<<"8BITMIME">>,true},
                {<<"AUTH">>,
                 <<"LOGIN PLAIN XOAUTH2 PLAIN-CLIENTTOKEN OAUTHBEARER XOAUTH">>},
                {<<"ENHANCEDSTATUSCODES">>,true},
                {<<"PIPELINING">>,true},
                {<<"CHUNKING">>,true},
                {<<"SMTPUTF8">>,true}]
Auth types: <<"LOGIN PLAIN XOAUTH2 PLAIN-CLIENTTOKEN OAUTHBEARER XOAUTH">>
Fixed types: ["LOGIN","PLAIN","XOAUTH2","PLAIN-CLIENTTOKEN","OAUTHBEARER",
              "XOAUTH"]
available authentication types, in order of preference: ["LOGIN","PLAIN"]
username prompt
password prompt
authentication accepted
Authentication status is true
Mail sending successful
[<<"2.0.0 OK 1485240254 a25sm2423382pgd.26 - gsmtp\r\n">>,
 <<"2.0.0 OK 1485240255 a25sm2423382pgd.26 - gsmtp\r\n">>]
```

Summary of changes:

* fixed a bunch of broken tests
* Fall back to A records if there's no MX record
* Add an optional trace function you can supply to print/log the progress of the SMTP session

I will finish cleaning things up (mainly to ensure backwards compatability) and make a PR for people to test.

**Comment by Vagabond** [comment]
see #117

**Comment by Shemeikka** [comment]
Hi,

Is this still a valid issue or could it be closed? The linked pull request (#117) does seem to allow sending multiple emails using just one TCP connection which should fix the problem @skosch was seeing.

**Comment by skosch** [comment]
Yes, @Vagabond solved this and I am using this in production – I should have closed this a long time ago!
